xonsh/environ.py

```diff
@@ -2250,23 +2249,41 @@ def foreign_env_fixes(ctx):
         del ctx["PROMPT"]
 
 
+class _RcPath(str):
+    """A class used exclusively to know which entry was added temporarily to
+    sys.path while loading rc files.
+    """
+
+    pass
+
+
 def xonsh_script_run_control(filename, ctx, env, execer=None, login=True):
     """Loads a xonsh file and applies it as a run control."""
+
     if execer is None:
         return False
+
     updates = {"__file__": filename, "__name__": os.path.abspath(filename)}
     try:
+        rc_dir = _RcPath(os.path.dirname(filename))
+        sys.path.append(rc_dir)
         with swap_values(ctx, updates):
             run_script_with_cache(filename, execer, ctx)
         loaded = True
+
     except SyntaxError as err:
         msg = "syntax error in xonsh run control file {0!r}: {1!s}"
         print_exception(msg.format(filename, err))
         loaded = False
+
     except Exception as err:
         msg = "error running xonsh run control file {0!r}: {1!s}"
         print_exception(msg.format(filename, err))
         loaded = False
+
+    finally:
+        sys.path = list(filter(lambda p: not isinstance(p, _RcPath), sys.path))
```

Comment:
Maybe you can use a set and check if each element is in that set instead of making a whole new type?

Reply:
The issue with this is that the user might have manually added the rc folder to `sys.path`.

If we want to leave the path as if we never added the rc folder (as we are currently doing here), then we need to be able to distinguish between the user added path and our own to avoid removing the manually added path, hence the wrapper type.

If we change the approach to permanently add the folder of any available rc files to path, then differentiating between our path and the user added path becomes a non-issue.

Reply:
I don't understand this either. What prevents you from keeping a copy of the original `sys.path` contents and restoring it after the RC has finished running?
Your solution clearly works, but it has quite a high WTF factor and is probably even slower than making a copy of the list.

Reply:
> Edit: To answer myself, that's probably in order to preserve the ability to modify sys.path from within the RC script, right?

Yeah I think that makes sense

Reply:
> Edit: To answer myself, that's probably in order to preserve the ability to modify sys.path from within the RC script, right?

Yeah I think that makes sense

Reply:
Hello! Sorry for the delayed response. Yes, that is intended to allow to edit `sys.path` inside rc files.